[Uid] Add the UidValueResolver argument value resolver

fancyweb opened the pull request.

guillaume-a committed:
[Console] Issue 43602 : Add fish completion

fancyweb committed:
[Uid] Add the UidValueResolver argument value resolver

jderusse committed:
Make RateLimiter resilient to timeShifting

Korbeil committed:
Fix aliased namespaces matching

sidz committed:
[Validator] Allow Sequence constraint to be applied onto class as an attribute

Commit:
[PropertyInfo] Add support for phpDocumentor and PHPStan pseudo-types

goodjinny committed:
[Workflow] Fix eventsToDispatch parameter setup for StateMachine

mostafazs committed:
[Form] Improve Persian (Farsi) Translation For Forms

alexandre-daubois committed:
[Config] In XmlUtils, avoid converting from octal every string starting with a 0

alexander-schranz committed:
Make enable_authenticator_manager true as there is no other way in Symfony 6

fancyweb committed:
[HttpClient] Allow yielding Exception from MockResponse's $body to mock transport errors

nicolas-grekas committed:
[HttpClient] Don't reset timeout counter when initializing requests

nicolas-grekas committed:
bug #44571 [HttpClient] Don't reset timeout counter when initializing requests (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[HttpClient] Don't reset timeout counter when initializing requests

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44544
| License       | MIT
| Doc PR        | -

Same as #42896 but on initialize.

Commits
-------

8f3bdeb359 [HttpClient] Don't reset timeout counter when initializing requests

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  [HttpClient] Don't reset timeout counter when initializing requests

GromNaN committed:
[Framework] Env var SYMFONY_IDE to set framework.ide

sbelyshkin committed:
[Cache] Fix saving items with no expiration through ProxyAdapter

wouterj committed:
bug #44554 Make enable_authenticator_manager true as there is no other way in Symfony 6 (alexander-schranz)

This PR was merged into the 6.0 branch.

Discussion
----------

Make enable_authenticator_manager true as there is no other way in Symfony 6

| Q             | A
| ------------- | ---
| Branch?       | 6.0 for features / 4.4, 5.3, 5.4 or 6.0 for bug fixes <!-- see below -->
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | /no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

Did read the new docs and I was confused about:

```
security:
    enable_authenticator_manager: true
```

as I thought thats the default now and there is no other way. And that seems to case as if you set it to false it will error in: https://github.com/symfony/symfony/blob/6ab662b0a8d5deabe03920a6dc0c7cbdc50a8d21/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L93-L95

So I would remove the `enable_authenticator_manager` also from the 6.0 docs and make this just the default to avod confusion here.

Commits
-------

b7a2ae6499 Make enable_authenticator_manager true as there is no other way in Symfony 6

fractalzombie committed:
[Messenger] Added generic template for `Envelope::last()` method

Nek- committed:
fix: lowest version of psr container supported

fancyweb committed:
[Uid] Add ulid keyword in composer.json

nicolas-grekas committed:
minor #44587 [Uid] Add ulid keyword in composer.json (fancyweb)

This PR was merged into the 5.3 branch.

Discussion
----------

[Uid] Add ulid keyword in composer.json

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Commits
-------

669b75f214 [Uid] Add ulid keyword in composer.json

derrabus committed:
minor #44586 [DependencyInjection] PSR Container dependency version fix for 5.3 & 5.4 (Nek-)

This PR was merged into the 5.3 branch.

Discussion
----------

[DependencyInjection] PSR Container dependency version fix for 5.3 & 5.4

| Q             | A
| ------------- | ---
| Branch?       | 5.3 / 5.4
| Bug fix?      | somehow
| New feature?  | no
| Deprecations? | no
| License       | MIT
<!--
Replace this notice by a short README for your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch 5.x.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

This problem is already solved on Sf 6.0 branch and also on the component itself but for 5.4 and 5.3 if you install with `--prefer-lowest` for `symfony/symfony` package, it will install a wrong version of psr container.

Updated from https://github.com/symfony/symfony/pull/44561 with `@fabpot` feedbacks.

`@derrabus` see the correct version here: https://github.com/symfony/symfony/blob/5.3/src/Symfony/Component/DependencyInjection/composer.json#L20

Commits
-------

b7644bd2aa fix: lowest version of psr container supported

wouterj committed:
minor #44591 [Form] Improve Persian (Farsi) Translation For Forms (mostafazs)

This PR was squashed before being merged into the 4.4 branch.

Discussion
----------

[Form] Improve Persian (Farsi) Translation For Forms

This is a 4.4 targeting PR of https://github.com/symfony/symfony/pull/44512

Original body:

> | Q             | A
> | ------------- | ---
> | Branch?       | 6.1
> | Bug fix?      | no
> | New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
> | Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
> | License       | MIT
>
> Improve Persian (Farsi) Translation For Form Component Translation File.

Commits
-------

f5c2f53eec [Form] Improve Persian (Farsi) Translation For Forms

nicolas-grekas committed:
[PropertyInfo] fix precedence of __get() vs properties

fancyweb commented on src/Symfony/Component/Uid/ArgumentResolver/UidValueResolver.php:
```diff
@@ -0,0 +1,71 @@
+        $uidClass = $argument->getType();
+
+        $c = match ($format) {
+            Uid::FORMAT_CANONICAL => static fn (string $uid): AbstractUid => new $uidClass($uid), // TODO: this is wrong
```
@nicolas-grekas Est-ce qu'on peut rajouter `abstract public static function fromCanonical(string $uid): static;` dans `AbstractUid`, je ne vois pas comment instancier les classes custom qui extend direct `AbstractUid`  au format canonical (sans le connaître) ? Le constructeur ne faisant pas partie du contrat de `AbstractUid`.

fancyweb replied:
Voir https://github.com/fancyweb/symfony/commit/fce7a6173481d0e8eb57247056d19e68ade9c9b7

nicolas-grekas replied:
check after the creating that __toString() === $uid?

fancyweb commented on src/Symfony/Component/Uid/ArgumentResolver/UidValueResolver.php:
```diff
@@ -0,0 +1,71 @@
+        return !$argument->isVariadic() &&
+            \is_string($request->attributes->get($argument->getName())) &&
+            null !== $argument->getType() &&
+            is_a($argument->getType(), AbstractUid::class, true);
```
is_subclass_of actually I think

fancyweb replied:
Only for fromString since it cannot be called on abstract class.

fancyweb replied:
+ not abstract class

kgonella committed:
[Translation] Handle the blank-translation in Loco Adapter

derrabus committed:
Make use of the nullsafe operator

fancyweb committed:
[Serializer] Fix denormalizing custom class in UidNormalizer

nicolas-grekas committed:
[HttpClient] Fix closing curl-multi handle too early on destruct

nicolas-grekas committed:
bug #44601 [HttpClient] Fix closing curl-multi handle too early on destruct (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[HttpClient] Fix closing curl-multi handle too early on destruct

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44334
| License       | MIT
| Doc PR        | -

For some reason, the garbage collector can decide to destruct the `CurlClientState` before the responses that reference them.
When this happens, the curl-multi handle is closed and responses end up in a broken state.
This fixes it by not closing the multi-handle on destruct/reset.

This also fixes configuring the multi-handle on reset.

Commits
-------

c0602fde4e [HttpClient] Fix closing curl-multi handle too early on destruct

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  [HttpClient] Fix closing curl-multi handle too early on destruct
  [Form] Improve Persian (Farsi) Translation For Forms

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  [HttpClient] Fix closing curl-multi handle too early on destruct
  [PropertyInfo] fix precedence of __get() vs properties
  [Form] Improve Persian (Farsi) Translation For Forms
  [Uid] Add ulid keyword in composer.json
  fix: lowest version of psr container supported
  [HttpClient] Don't reset timeout counter when initializing requests

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  [HttpClient] Fix closing curl-multi handle too early on destruct
  [PropertyInfo] fix precedence of __get() vs properties
  [Form] Improve Persian (Farsi) Translation For Forms
  [Uid] Add ulid keyword in composer.json
  fix: lowest version of psr container supported
  [HttpClient] Don't reset timeout counter when initializing requests

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  [HttpClient] Fix closing curl-multi handle too early on destruct
  [PropertyInfo] fix precedence of __get() vs properties
  [Form] Improve Persian (Farsi) Translation For Forms
  [Uid] Add ulid keyword in composer.json
  fix: lowest version of psr container supported
  Make enable_authenticator_manager true as there is no other way in Symfony 6
  [HttpClient] Don't reset timeout counter when initializing requests

nicolas-grekas committed:
[DI] fix merge

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  [DI] fix merge

fabpot committed:
feature #43641 [Console] Issue 43602 : Add fish completion (guillaume-a)

This PR was squashed before being merged into the 6.1 branch.

Discussion
----------

[Console] Issue 43602 : Add fish completion

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #43602
| License       | MIT

I tried to implement fish auto completion.
I hope it will work fine for you guys, let me know.

You can execute `bin/console completion fish >> ~/.config/fish/completions/sf_console.fish` and source the file

Commits
-------

157c8c0cb0 [Console] Issue 43602 : Add fish completion

GromNaN committed:
feature #44575 [Framework] Read env var SYMFONY_IDE by default for framework.ide (GromNaN)

This PR was merged into the 6.1 branch.

Discussion
----------

[Framework] Read env var SYMFONY_IDE by default for framework.ide

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Choosing an IDE is a personal preference, it should not be a project configuration ([doc](https://symfony.com/doc/current/reference/configuration/framework.html#ide)). In a team, each developer can use a different IDE, and a developer generally uses the same IDE on all their projects.

To setup a system-wide config, it is possible to update `xdebug.file_link_format` in php.ini. But that have to be done on every php version installed or any docker image; and we miss the shortcuts provided by Symfony.

My proposition it to read an environment variable that can be set by the developers in their user profile or in `.env.local`. The name `SYMFONY_IDE` can be discussed.

I currently uses the following config:
```yaml
framework:
    ide: "%env(string:default::SYMFONY_IDE)%"
```

Commits
-------

4d4f8b470a [Framework] Env var SYMFONY_IDE to set framework.ide

fabpot committed:
minor #44597 Make use of the nullsafe operator (derrabus)

This PR was merged into the 6.1 branch.

Discussion
----------

Make use of the nullsafe operator

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

PHP 8 introduced a nullsafe operator which allows us to simplify some of our code.

I have used a PhpStorm inspection to automate the change. I did some minor adjustments afterwards where the fixer left unnecessary paranthesis or local variables.

Commits
-------

d58449db23 Make use of the nullsafe operator

fancyweb commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -2037,6 +2037,15 @@ private function addUidSection(ArrayNodeDefinition $rootNode, callable $enableIf
                         ->scalarNode('time_based_uuid_node')
                             ->cannotBeEmpty()
                         ->end()
+                        ->arrayNode('argument_value_resolver')
```
TODO: xml config

nicolas-grekas committed:
bug #44430 [PropertyInfo] Fix aliased namespace matching (Korbeil)

This PR was merged into the 5.4 branch.

Discussion
----------

[PropertyInfo] Fix aliased namespace matching

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44417
| License       | MIT
| Doc PR        | N/A

When you have an aliased namespace that you use in your code it was erroring because we didn't stored the first part of the namespace that should be replaced.

Commits
-------

432b1a14a5 Fix aliased namespaces matching

nicolas-grekas committed:
[HttpClient] minor change

nicolas-grekas committed:
[HttpClient] fix monitoring responses issued before reset()

nicolas-grekas committed:
[HttpClient] Fix dealing with "HTTP/1.1 000 " responses

nicolas-grekas committed:
[Cache] workaround transient test on M1

fancyweb committed:
[Intl] Leverage array_is_list

nicolas-grekas committed:
minor #44624 [HttpClient] workaround transient test on M1 (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[HttpClient] workaround transient test on M1

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Commits
-------

54efeab581 [Cache] workaround transient test on M1

nicolas-grekas committed:
bug #44623 [HttpClient] Fix dealing with "HTTP/1.1 000 " responses (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[HttpClient] Fix dealing with "HTTP/1.1 000 " responses

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44617
| License       | MIT
| Doc PR        | -

Commits
-------

d793d03cb6 [HttpClient] Fix dealing with "HTTP/1.1 000 " responses

nicolas-grekas committed:
bug #44625 [HttpClient] fix monitoring responses issued before reset() (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[HttpClient] fix monitoring responses issued before reset()

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Follow up of #44601

In order to reset the DNS cache, we have to create a new curl-multi handle. But since there might still be responses attached to the previous handle, we need to keep it around to monitor it.

Best reviewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/44625/files?w=1).

Commits
-------

6e529608c0 [HttpClient] fix monitoring responses issued before reset()

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  [HttpClient] fix monitoring responses issued before reset()
  [Cache] workaround transient test on M1
  [HttpClient] Fix dealing with "HTTP/1.1 000 " responses
  [HttpClient] minor change

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  [HttpClient] fix monitoring responses issued before reset()
  [Cache] workaround transient test on M1
  [HttpClient] Fix dealing with "HTTP/1.1 000 " responses
  [HttpClient] minor change

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  [HttpClient] fix monitoring responses issued before reset()
  [Cache] workaround transient test on M1
  [HttpClient] Fix dealing with "HTTP/1.1 000 " responses
  [HttpClient] minor change
  Fix aliased namespaces matching

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  [HttpClient] fix monitoring responses issued before reset()
  [Cache] workaround transient test on M1
  [HttpClient] Fix dealing with "HTTP/1.1 000 " responses
  [HttpClient] minor change
  Fix aliased namespaces matching

derrabus committed:
[Workflow] Add return type annotations to the Event class

fabpot committed:
minor #44626 [Intl] Leverage array_is_list (fancyweb)

This PR was merged into the 6.1 branch.

Discussion
----------

[Intl] Leverage array_is_list

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | -
| Tickets       | https://github.com/symfony/symfony/pull/44167#pullrequestreview-811851606
| License       | MIT
| Doc PR        | -

It was forgotten in https://github.com/symfony/symfony/pull/44167 and this is the last occurrence in the code base :pray:

Commits
-------

575ebf294f [Intl] Leverage array_is_list

fabpot committed:
bug #44510 [Workflow] Fix eventsToDispatch parameter setup for StateMachine (Olexandr Kalaidzhy)

This PR was merged into the 5.3 branch.

Discussion
----------

[Workflow] Fix eventsToDispatch parameter setup for StateMachine

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | see description
| License       | MIT
| Doc PR        | N/A

Using `state_machine` type of the workflow in the project have mentioned that workflow events still propagated with `events_to_dispatch: []` parameter. But this bug does not appear for `workflow` type.

Current pull request fixes that bug by adding $eventsToDispatch parameter initialization for StateMachine class.

Before fixing (`state_machine` with `events_to_dispatch: []`):
![Screenshot from 2021-12-08 10-55-51](https://user-images.githubusercontent.com/13361839/145183068-e419b48e-129d-424a-89d4-5c2ed77b0695.png)

After fixing (`state_machine` with `events_to_dispatch: []`);
![Screenshot from 2021-12-08 10-54-58](https://user-images.githubusercontent.com/13361839/145183132-f227d927-f385-4ede-b60d-87d1f8fa2bf4.png)

Commits
-------

ca38501915 [Workflow] Fix eventsToDispatch parameter setup for StateMachine

fabpot committed:
bug #44537 [Config] In XmlUtils, avoid converting from octal every string starting with a 0 (alexandre-daubois)

This PR was merged into the 4.4 branch.

Discussion
----------

[Config] In XmlUtils, avoid converting from octal every string starting with a 0

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #41532 <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | N/A

One word on the use of `intval` instead of `octdec`: `octdec` leads to deprecation notices when testing a non-octal string:
```
Invalid characters passed for attempted conversion, these have been ignored
```
While `intval` doesn't, which worth the use in my opinion.

Commits
-------

30c3913eb2 [Config] In XmlUtils, avoid converting from octal every string starting with a 0

nicolas-grekas committed:
[HttpClient] fix segfault when canary is triggered after the curl handle is destructed

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  [HttpClient] fix segfault when canary is triggered after the curl handle is destructed
  [Workflow] Fix eventsToDispatch parameter setup for StateMachine

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  [HttpClient] fix segfault when canary is triggered after the curl handle is destructed
  [Workflow] Fix eventsToDispatch parameter setup for StateMachine

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  [HttpClient] fix segfault when canary is triggered after the curl handle is destructed
  [Workflow] Fix eventsToDispatch parameter setup for StateMachine

nicolas-grekas committed:
Fix CI on macos-11

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  Fix CI on macos-11
  [Config] In XmlUtils, avoid converting from octal every string starting with a 0

nicolas-grekas committed:
Try making tests a bit less transient

nicolas-grekas committed:
minor #44636 Try making tests a bit less transient (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

Try making tests a bit less transient

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Commits
-------

b1426b4df8 Try making tests a bit less transient

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  Try making tests a bit less transient

nicolas-grekas committed:
bug #44600 [Serializer] Fix denormalizing custom class in UidNormalizer (fancyweb)

This PR was merged into the 5.3 branch.

Discussion
----------

[Serializer] Fix denormalizing custom class in UidNormalizer

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Considering we support denormalizing any subclass of `AbstractUid::class` we should not denormalize any `$type !== Ulid::class` to `Uuid::class` because this is wrong. We need to call `$type::fromString` (see `TestCustomUid`).
But `fromString()` is an abstract method, so it cannot be called directly on abstract classes, I guess we should have not supported them from the start. However, since they work right now (since all abstract classes fall back to the `Uuid:class` behavior) we cannot break it. So I propose to deprecate supporting abstract classes in 6.1 and remove it in 7.0.

Commits
-------

ba9e0028ef [Serializer] Fix denormalizing custom class in UidNormalizer

nicolas-grekas committed:
bug #44354 [RateLimiter] Make RateLimiter resilient to timeShifting (jderusse)

This PR was merged into the 5.3 branch.

Discussion
----------

[RateLimiter] Make RateLimiter resilient to timeShifting

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

When 2 servers use the FixedWindowLimiter and do not have the exact same clock (with microsecond precision) the `Window` might return 0 available tokens.

Commits
-------

6918a21a3b Make RateLimiter resilient to timeShifting

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  Try making tests a bit less transient
  Fix CI on macos-11
  [Serializer] Fix denormalizing custom class in UidNormalizer
  [Config] In XmlUtils, avoid converting from octal every string starting with a 0
  Make RateLimiter resilient to timeShifting

nicolas-grekas committed:
bug #44448 [Validator] Allow Sequence constraint to be applied onto class as an attribute (sidz)

This PR was squashed before being merged into the 5.4 branch.

Discussion
----------

[Validator] Allow Sequence constraint to be applied onto class as an attribute

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/issues/44531
| License       | MIT
| Doc PR        | N/A

This PR allows to apply `Sequentially` constraint as an attribute onto class target.

Looks like it was missed on translating annotations to attributes.

![image](https://user-images.githubusercontent.com/1302230/144623891-98bde6b2-2376-4ac0-93fc-58b43a0b0f94.png)
![image](https://user-images.githubusercontent.com/1302230/144623958-7d5fe384-771a-4711-8c28-0e9bddfc4c41.png)

Commits
-------

8c1e746a65 [Validator] Allow Sequence constraint to be applied onto class as an attribute

derrabus committed:
[Workflow] Remove redundant type check

derrabus committed:
Merge branch '5.4' into 6.0

* 5.4:
  [Validator] Allow Sequence constraint to be applied onto class as an attribute
  Try making tests a bit less transient
  Fix CI on macos-11
  [Serializer] Fix denormalizing custom class in UidNormalizer
  [Config] In XmlUtils, avoid converting from octal every string starting with a 0
  Make RateLimiter resilient to timeShifting

derrabus committed:
Merge branch '6.0' into 6.1

* 6.0:
  [Validator] Allow Sequence constraint to be applied onto class as an attribute
  Try making tests a bit less transient
  Fix CI on macos-11
  [Serializer] Fix denormalizing custom class in UidNormalizer
  [Config] In XmlUtils, avoid converting from octal every string starting with a 0
  Make RateLimiter resilient to timeShifting

fancyweb committed:
[Serializer] Fix symfony/uid requirement

lyrixx committed:
[Workflow] Fix Event constructor requirements

This has already been deprecied in 4.x
But while cleaning the 5.x branch, I forgot to apply this patch

See
https://github.com/symfony/symfony/pull/31824/files#diff-5f386ffb0109cc731bd98e63eea021b32faadd98791bd6ba65926d09c5e2ec40L37

nicolas-grekas committed:
Skip transient tests on macos

nicolas-grekas committed:
minor #44643 Skip transient tests on macos (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

Skip transient tests on macos

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Commits
-------

880988cb0b Skip transient tests on macos

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  Skip transient tests on macos

nicolas-grekas committed:
minor #44641 [Serializer] Fix symfony/uid requirement (fancyweb)

This PR was merged into the 5.3 branch.

Discussion
----------

[Serializer] Fix symfony/uid requirement

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | -
| Tickets       | -
| License       | MIT
| Doc PR        | -

https://github.com/symfony/symfony/pull/44600 needs https://github.com/symfony/symfony/commit/824777938db7066734f6e18ddbaccb4f69bc8318

Commits
-------

6a0fd54987 [Serializer] Fix symfony/uid requirement

nicolas-grekas committed:
bug #44598 [Translation] Handle the blank-translation in Loco Adapter (kgonella)

This PR was merged into the 5.3 branch.

Discussion
----------

[Translation] Handle the blank-translation in Loco Adapter

| Q             | A
| ------------- | ---
| Branch?       | 5.3, 5.4, 6.0
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44595
| License       | MIT

This PR aim to be able to pull the "blank-translation" in addition to "translated" status translations. See the ticket to know more about. I don't think we can add a unit test here as the calls to loco are fully mocked.

Commits
-------

8b18a273df [Translation] Handle the blank-translation in Loco Adapter

nicolas-grekas committed:
minor #44642 [Workflow] Fix Event constructor requirements (lyrixx)

This PR was merged into the 5.3 branch.

Discussion
----------

[Workflow] Fix Event constructor requirements

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

This has already been deprecied in 4.x
But while cleaning the 5.x branch, I forgot to apply this patch

See
https://github.com/symfony/symfony/pull/31824/files#diff-5f386ffb0109cc731bd98e63eea021b32faadd98791bd6ba65926d09c5e2ec40L37

More over, `Event::getWorkflow()` has the correct return type.

Commits
-------

acc1cdb431 [Workflow] Fix Event constructor requirements

nicolas-grekas committed:
Revert "minor #44642 [Workflow] Fix Event constructor requirements (lyrixx)"

This reverts commit 18c7edd2e481c4043a2552873ee0f62395e2b367, reversing
changes made to e5d3deaf2edf469bc4c542d00d43428fa8d33dc2.

nicolas-grekas committed:
minor #44638 [Workflow] Remove redundant type check (derrabus)

This PR was merged into the 6.0 branch.

Discussion
----------

[Workflow] Remove redundant type check

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

`MarkingStoreInterface::getMarking()` has a native return type `Marking`. The type check I've removed here is thus 100% redundant and can be removed.

Commits
-------

60b1672834 [Workflow] Remove redundant type check

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  Revert "minor #44642 [Workflow] Fix Event constructor requirements (lyrixx)"
  Skip transient tests on macos
  [Workflow] Fix Event constructor requirements
  [Translation] Handle the blank-translation in Loco Adapter
  [Serializer] Fix symfony/uid requirement

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  Revert "minor #44642 [Workflow] Fix Event constructor requirements (lyrixx)"
  Skip transient tests on macos
  [Workflow] Fix Event constructor requirements
  [Translation] Handle the blank-translation in Loco Adapter
  [Serializer] Fix symfony/uid requirement

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  Revert "minor #44642 [Workflow] Fix Event constructor requirements (lyrixx)"
  [Workflow] Remove redundant type check
  Skip transient tests on macos
  [Workflow] Fix Event constructor requirements
  [Translation] Handle the blank-translation in Loco Adapter
  [Serializer] Fix symfony/uid requirement

nicolas-grekas committed:
[VarDumper] add more "transient-on-macos" groups

nicolas-grekas committed:
[HttpKernel] fix how configuring log-level and status-code by exception works

nicolas-grekas committed:
Add more nullsafe operators

nicolas-grekas committed:
Revert "feature #41989 [Cache] make `LockRegistry` use semaphores when possible (nicolas-grekas)"

This reverts commit 479919d4d5cbc17710d16a04ff9728d1519afb8e, reversing
changes made to 356c9533cc520b5aab5d92d0b06a408beb22a157.

nicolas-grekas committed:
[Cache] disable lock on CLI

sylfabre committed:
[Translation] Translatable parameters

nicolas-grekas committed:
[HttpClient] Fix tracing requests made after calling withOptions()

derrabus committed:
Remove pointless assignment

nicolas-grekas committed:
minor #44674 [Messenger] Remove pointless assignment (derrabus)

This PR was merged into the 6.0 branch.

Discussion
----------

[Messenger] Remove pointless assignment

| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

Commits
-------

981cb0c2b7 Remove pointless assignment

nicolas-grekas committed:
fix merge

nicolas-grekas committed:
CS fixes

nicolas-grekas committed:
minor #44679 CS fixes (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

CS fixes

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Running php-cs-fixer on the codebase

Commits
-------

6e3b71125a CS fixes

nicolas-grekas committed:
bug #44669 [Cache] disable lock on CLI (nicolas-grekas)

This PR was merged into the 4.4 branch.

Discussion
----------

[Cache] disable lock on CLI

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? |
| Tickets       | Fix #41130
| License       | MIT
| Doc PR        | -

I thought using semaphores would fix the linked issue but I'm proposing to revert them in #44667

Instead, I think we should disable the lock on CLI processes.

The purpose of having locks enabled by default is to prevent cache stamped at deploy time. Disabling the lock on CLI still provides this protection on eg FPM.

For ppl that really want a lock between eg crons and FPM, the lock component should be used instead.

Commits
-------

b968514286 [Cache] disable lock on CLI

nicolas-grekas committed:
bug #44577 [Cache] Fix proxy no expiration to the Redis (Sergey Belyshkin)

This PR was merged into the 4.4 branch.

Discussion
----------

[Cache] Fix proxy no expiration to the Redis

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | - <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        |

After #38156 RedisAdapter is not able to persist items with no expiration which come through ProxyAdapter because Redis does not persist keys with very large TTL such as \PHP_INT_MAX (64 bits).

Commits
-------

0812726975 [Cache] Fix saving items with no expiration through ProxyAdapter

nicolas-grekas committed:
Merge branch '4.4' into 5.3

* 4.4:
  [Cache] Fix saving items with no expiration through ProxyAdapter
  CS fixes
  [Cache] disable lock on CLI
  [VarDumper] add more "transient-on-macos" groups

nicolas-grekas committed:
[5.3] cs fixes

nicolas-grekas committed:
bug #44671 [HttpClient] Fix tracing requests made after calling withOptions() (nicolas-grekas)

This PR was merged into the 5.3 branch.

Discussion
----------

[HttpClient] Fix tracing requests made after calling withOptions()

| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44312
| License       | MIT
| Doc PR        | -

Commits
-------

06b25c713b [HttpClient] Fix tracing requests made after calling withOptions()

nicolas-grekas committed:
Merge branch '5.3' into 5.4

* 5.3:
  [5.3] cs fixes
  [Cache] Fix saving items with no expiration through ProxyAdapter
  CS fixes
  [HttpClient] Fix tracing requests made after calling withOptions()
  [Cache] disable lock on CLI
  [VarDumper] add more "transient-on-macos" groups

nicolas-grekas committed:
[5.4] cs fixes

nicolas-grekas committed:
bug #44667 [Cache] Revert "feature #41989  make `LockRegistry` use semaphores when possible" (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[Cache] Revert "feature #41989  make `LockRegistry` use semaphores when possible"

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44536
| License       | MIT
| Doc PR        | -

This reverts commit 479919d4d5cbc17710d16a04ff9728d1519afb8e, reversing
changes made to 356c9533cc520b5aab5d92d0b06a408beb22a157.

I'd like to revert this PR because using semaphores is creating problems that don't have any easy solution.

Commits
-------

6bcc3cb2de Revert "feature #41989 [Cache] make `LockRegistry` use semaphores when possible (nicolas-grekas)"

nicolas-grekas committed:
bug #44649 [HttpKernel] fix how configuring log-level and status-code by exception works (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[HttpKernel] fix how configuring log-level and status-code by exception works

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44433
| License       | MIT
| Doc PR        | -

This PR replaces #44456, and essentially fixes the way configuring log-level/status-code by exception works: instead of hacking them in the listener, the original exception should be wrapped in an `HttpExceptionInterface`, so that the rest of the code can seamlessly know about the configuration.

Commits
-------

6617be449e [HttpKernel] fix how configuring log-level and status-code by exception works

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  [5.4] cs fixes
  [5.3] cs fixes
  [Cache] Fix saving items with no expiration through ProxyAdapter
  CS fixes
  [HttpClient] Fix tracing requests made after calling withOptions()
  [Cache] disable lock on CLI
  Revert "feature #41989 [Cache] make `LockRegistry` use semaphores when possible (nicolas-grekas)"
  [HttpKernel] fix how configuring log-level and status-code by exception works
  [VarDumper] add more "transient-on-macos" groups

nicolas-grekas committed:
[6.0] cs fixes

nicolas-grekas committed:
Merge branch '6.0' into 6.1

* 6.0:
  [6.0] cs fixes
  [5.4] cs fixes
  [5.3] cs fixes
  [Cache] Fix saving items with no expiration through ProxyAdapter
  CS fixes
  fix merge
  Remove pointless assignment
  [HttpClient] Fix tracing requests made after calling withOptions()
  [Cache] disable lock on CLI
  Revert "feature #41989 [Cache] make `LockRegistry` use semaphores when possible (nicolas-grekas)"
  [HttpKernel] fix how configuring log-level and status-code by exception works
  [VarDumper] add more "transient-on-macos" groups

nicolas-grekas committed:
minor #44659 Add more nullsafe operators (nicolas-grekas)

This PR was merged into the 6.1 branch.

Discussion
----------

Add more nullsafe operators

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Commits
-------

56f908a1e1 Add more nullsafe operators

nicolas-grekas committed:
[6.1] cs fixes

nicolas-grekas committed:
minor #44583 [Messenger] Added generic template for `Envelope::last()` method (fractalzombie)

This PR was merged into the 6.1 branch.

Discussion
----------

[Messenger] Added generic template for `Envelope::last()` method

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| License       | MIT

Added possibility to PHPStorm generic support for Envelope::last method, no need more `/** `@var` SomeStampClass */`

Commits
-------

bbdc403c19 [Messenger] Added generic template for `Envelope::last()` method

nicolas-grekas committed:
feature #44451 [PropertyInfo] Add support for phpDocumentor and PHPStan pseudo-types (EmilMassey)

This PR was squashed before being merged into the 6.1 branch.

Discussion
----------

[PropertyInfo] Add support for phpDocumentor and PHPStan pseudo-types

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

More and more apps are using pseudo-types like `non-empty-string`, `positive-int` which are understood by static analysis tools but are not a part of the language. This PR adds support for these types to `PhpDocExtractor` and `PhpStanExtractor`. Pseudo-type is mapped to built-in type(s) (e.g. `non-empty-string` => `string`, `positive-int` => `int`, `number` => `int|float`).

This PR adds support for all pseudo-types defined by the [phpDocumentor](https://github.com/phpDocumentor/TypeResolver/tree/f8ec4ab631de5a97769e66b13418c3b8b24e81f4/src/PseudoTypes) (some of them like `list` or `true`, `false` are already supported) and [PHPStan's](https://phpstan.org/writing-php-code/phpdoc-types) basic types.

Commits
-------

3e49b0f002 [PropertyInfo] Add support for phpDocumentor and PHPStan pseudo-types

nicolas-grekas committed:
Sync expected missing return types

nicolas-grekas committed:
Skip patching Envelope's return type

nicolas-grekas committed:
Fix patching return type of Envelope

fabpot committed:
feature #44672 [Translation] Translatable parameters (sylfabre)

This PR was merged into the 6.1 branch.

Discussion
----------

[Translation] Translatable parameters

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | TBD

✅  The TwigExtension now supports messages implementing `TranslatableInterface` (https://symfony.com/blog/new-in-symfony-5-2-translatable-objects).
✅ Thanks to #41858, `TranslatableMessage` now supports recursive `TranslatableInterface` as params.

❌ But using `TranslatableInterface` as params with regular messages is not supported yet.

💡 This PR addresses this issue and makes it possible to create dedicated `TranslatableInterface` implementations like the one from https://github.com/symfony/symfony/pull/41136

_Note: I would like also to add `TranslatableInterface` to the `trans(?string $id)` signature method like with the one from the [TwigExtension](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Bridge/Twig/Extension/TranslationExtension.php#L111) but I don't know how to do it without breaking BC. Any advice is welcome!_

Commits
-------

5beaee85f9 [Translation] Translatable parameters

fabpot committed:
minor #44631 [Workflow] Add return type annotations to the Event class (derrabus)

This PR was merged into the 6.1 branch.

Discussion
----------

[Workflow] Add return type annotations to the Event class

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

Paving the road to a fully-typed Symfony 7. 😎

Commits
-------

ccc918655f [Workflow] Add return type annotations to the Event class

fancyweb committed:
[FrameworkBundle] Add @return TestContainer to KernelTestCase::getContainer()

fancyweb committed:
minor #44695 [FrameworkBundle] Add `@return` TestContainer to KernelTestCase::getContainer() (fancyweb)

This PR was merged into the 6.1 branch.

Discussion
----------

[FrameworkBundle] Add `@return` TestContainer to KernelTestCase::getContainer()

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

From what I understood, `test.service_container` can only be an instance of `TestContainer`. Changing the return type allows autocompletion for useful methods such as `getServiceIds()` and `getRemovedIds()` (I didn't understand why they didn't showed up in my test). That's a DX improvement, can it be applied on a lower branch?

Commits
-------

812619cec5 [FrameworkBundle] Add `@return` TestContainer to KernelTestCase::getContainer()

fancyweb committed:
feature #44568 [HttpClient] Allow yielding Exception from MockResponse's  $body to mock transport errors (fancyweb)

This PR was merged into the 6.1 branch.

Discussion
----------

[HttpClient] Allow yielding Exception from MockResponse's  $body to mock transport errors

| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

A minor DX feature to allow simulating error chunks more easily (instead of being forced to throw from a \Generator).

Commits
-------

bbfe9d8215 [HttpClient] Allow yielding Exception from MockResponse's $body to mock transport errors
